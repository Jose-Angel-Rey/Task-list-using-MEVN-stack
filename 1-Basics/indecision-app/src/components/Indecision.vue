<template>
  <section class="input-container">
    <input
      @keypress.enter="cleanInput"
      class="input"
      placeholder="Type your question here!"
      required
      title="Type your question here"
      type="text"
      v-model.trim="question"
    />
    <strong class="input__advice">* remember to finish with "?"</strong>
  </section>

  <template v-if="showAnswer && answer">
    <img :src="answer.image" :alt="answer.response" class="answer__image" />
    <section class="answer">
      <h2 class="answer__question">{{ question }}</h2>
      <h3 class="answer__title">{{ answer.response }}!</h3>
    </section>
  </template>
  <template v-else-if="showAnswer && !answer">
    <section class="answer">
      <h2 class="answer__title">Thinking...</h2>
    </section>
  </template>
</template>

<script>
export default {
  name: "Indecision",
  data() {
    return {
      answer: null,
      showAnswer: null,
      question: "",
    };
  },
  methods: {
    cleanInput() {
      this.question = "";
    },
    async getAnswer() {
      const response = await fetch("https://yesno.wtf/api");
      const data = await response.json();
      console.log("data", data);
      this.answer = {
        response: data.answer,
        image: data.image,
      };
      this.showAnswer = true;
    },
  },
  watch: {
    question(value) {
      this.showAnswer = false;
      if (value.includes("?")) {
        this.answer = null;
        this.showAnswer = true;
        this.getAnswer();
      }
    },
  },
};
</script>

<style lang="scss" scoped>
.answer {
  margin-top: 3rem;
  max-width: 50rem;
  position: relative;
  width: 90%;
  z-index: 100;

  &__question {
    color: #bdbdbd;
    font-size: 2rem;
    font-weight: bold;
  }

  &__title {
    color: #fff;
    font-size: 1.5rem;
    font-weight: bold;
    margin: 1.5rem auto;
    text-transform: capitalize;
  }

  &__image {
    height: 100%;
    left: 0;
    -o-object-fit: cover;
    object-fit: cover;
    -o-object-position: center;
    object-position: center;
    position: absolute;
    top: 0;
    width: 100%;
    z-index: -10;
  }
}

.input {
  border-radius: 0.5rem;
  border: 1px solid #ddd;
  font-size: 1rem;
  max-width: 50rem;
  outline: none;
  padding: 1rem;
  transition: all 0.5s;
  width: 100%;

  &-container {
    align-items: center;
    display: flex;
    flex-direction: column;
    max-width: 50rem;
    position: relative;
    width: 90%;
    z-index: 100;
  }

  &__advice {
    color: #bdbdbd;
    font-size: 0.9rem;
    margin-top: 1rem;
  }
}
</style>
