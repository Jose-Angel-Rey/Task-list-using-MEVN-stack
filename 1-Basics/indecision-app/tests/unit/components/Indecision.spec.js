import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision.vue";

describe("Indecision", () => {
  let wrapper;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    jest.clearAllMocks();
  });

  test("should match the snapShot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should not execute the getAnswer function if the '?' symbol has not been entered in the input", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    let input = wrapper.find("input");

    await input.setValue("Question");

    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  test("the getAnswer function must be executed when entering the '?' symbol in the input", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    let input = wrapper.find("input");

    await input.setValue("Question ?");

    expect(getAnswerSpy).toHaveBeenCalled();
  });

  test("getAnswer should return an image and text in response", async () => {
    await wrapper.vm.getAnswer();

    const answer = wrapper.vm.answer.response;
    const image = wrapper.find("img");

    expect(answer).toBe("yes");
    expect(image.exists()).toBeTruthy();
  });

  test("getAnswer must return null if the API call fails", async () => {
    const answer = wrapper.vm.answer;

    await wrapper.vm.getAnswer();
    fetch.mockImplementationOnce(() => Promise.reject(new Error("Error")));

    expect(answer).toBe(null);
  });
});
