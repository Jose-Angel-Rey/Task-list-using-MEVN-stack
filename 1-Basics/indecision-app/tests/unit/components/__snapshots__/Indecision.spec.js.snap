// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision should match the snapShot 1`] = `
<section class="input-container"><input class="input" placeholder="Type your question here!" required="" title="Type your question here" type="text"><strong class="input__advice">* remember to finish with "?"</strong></section>
<!--v-if-->
`;
